internal/lint/check_message_names_camel_case.go

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package lint
+
+import (
+	"github.com/emicklei/proto"
+	"github.com/uber/prototool/internal/strs"
+	"github.com/uber/prototool/internal/text"
+)
+
+var messageNamesCamelCaseChecker = NewAddChecker(
+	"MESSAGE_NAMES_CAMEL_CASE",
+	"Verifies that all non-extended message names are CamelCase.",
+	checkMessageNamesCamelCase,
+)
+
+func checkMessageNamesCamelCase(add func(*text.Failure), dirPath string, descriptors []*proto.Proto) error {
+	return runVisitor(messageNamesCamelCaseVisitor{baseAddVisitor: newBaseAddVisitor(add)}, descriptors)
+}
+
+type messageNamesCamelCaseVisitor struct {
+	baseAddVisitor
+}
+
+func (v messageNamesCamelCaseVisitor) VisitMessage(message *proto.Message) {
+	// for nested messages
+	for _, child := range message.Elements {
+		child.Accept(v)
+	}
+	if message.IsExtend {
```

Comment:
We lint for proto3, but support proto2 in Prototool.

Reply:
yeah, i'm of the same mindset as @amckinney, i'd rather keep this tool as 'forward looking' as possible, even if it is fairly trivial to support some proto2 functionality

Reply:
So I think there's a bit of misunderstanding here - proto2 isn't something just of the past, we basically have to support the syntax. Even if we only use protobuf 3.0.0 and above, proto2 files are all over the place. For example, the Well-Known Types https://github.com/google/protobuf/blob/master/src/google/protobuf/descriptor.proto are proto2. Instead of thinking it as a breaking version of the language, it's more like "we only use go 1.10 and above to build prototool, but you can use elements of go 1.4" (bad example but).

Reply:
gotcha

Reply:
Ah ok makes sense. Thanks for the clarification.